stats/stats_linux.go

```diff
@@ -0,0 +1,159 @@
+func New(ctx context.Context) *Stats {
+	s := new(Stats)
+
+	count, err := getCPUCount()
+	if err != nil {
+		slog.Info("failed to get cpu count", "err", err)
+	}
+
+	s.cpuCount = count
+
+	go func() {
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				s.tick()
+			case <-ctx.Done():
+			}
+		}
+	}()
+
+	return s
+
+}
+
+func (s *Stats) tick() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	loadAvg1Min, loadAvg5Min, loadAvg15Min, err := getLoadAvgInfo()
+	if err != nil {
+		slog.Info("failed to get loadavg", "err", err)
+	}
+
+	s.loadAvg1Min = loadAvg1Min
+	s.loadAvg5Min = loadAvg5Min
+	s.loadAvg15Min = loadAvg15Min
+
+	memTotal, memFree, memAvailable, err := getMemInfo()
+	if err != nil {
+		slog.Info("failed to get meminfo", "err", err)
+	}
+
+	s.memTotal = memTotal
+	s.memFree = memFree
+	s.memAvailable = memAvailable
+}
+
+func (s *Stats) GetStatsTags() map[string]string {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return map[string]string{
+		"subtrace_linux_cpu_count":             fmt.Sprintf("%d", s.cpuCount),
+		"subtrace_linux_loadavg_1_min":         s.loadAvg1Min,
+		"subtrace_linux_loadavg_5_min":         s.loadAvg5Min,
+		"subtrace_linux_loadavg_15_min":        s.loadAvg15Min,
+		"subtrace_linux_meminfo_mem_total":     s.memTotal,
+		"subtrace_linux_meminfo_mem_free":      s.memFree,
+		"subtrace_linux_meminfo_mem_available": s.memAvailable,
+	}
+}
+
+func getCPUCount() (int, error) {
+	b, err := os.ReadFile("/proc/cpuinfo")
+	if err != nil {
+		return 0, fmt.Errorf("read /proc/cpuinfo: %w", err)
+	}
+
+	blocks := strings.Split(string(b), "\n\n")
+	return len(blocks) - 1, nil
+}
+
+// From [1]:
+//
+// /proc/loadavg
+// The first three fields in this file are load average figures giving the number
+// of jobs in the run queue (state R) or waiting for disk I/O (state D) averaged
+// over 1, 5, and 15 minutes.
+//
+// [1] https://man7.org/linux/man-pages/man5/proc_loadavg.5.html
+func getLoadAvgInfo() (string, string, string, error) {
+	b, err := os.ReadFile("/proc/loadavg")
+	if err != nil {
+		return "", "", "", fmt.Errorf("read /proc/loadavg: %w", err)
+	}
+
+	values := strings.Split(string(b), " ")
+	if len(values) < 3 {
+		return "", "", "", fmt.Errorf("not enough info in /proc/loadavg, expected at least 3 values but got %d", len(values))
+	}
+
+	return values[0], values[1], values[2], nil
+
+}
+
+// See [1] for details about the /proc/meminfo file.
+//
+// [1] https://man7.org/linux/man-pages/man5/proc_meminfo.5.html
+func getMemInfo() (string, string, string, error) {
+	b, err := os.ReadFile("/proc/meminfo")
+	if err != nil {
+		return "", "", "", fmt.Errorf("read: %w", err)
+	}
+
+	lines := strings.Split(string(b), "\n")
+	if len(lines) < 3 {
+		return "", "", "", fmt.Errorf("not enough info, expected at least 3 lines but got %d", len(lines))
+	}
+
+	var memTotal, memFree, memAvailable string
+	for i := 0; i < 3; i++ {
+		words := strings.Split(lines[i], " ")
```

Comment:
you can use `strings.Fields` to do this much more cleanly. it splits based on arbitrary length whitespace, not just a single space character (in `/proc/meminfo`, the RHS is padded with whitespace).